src/main/java/teammates/logic/core/EmailSender.java

```diff
@@ -0,0 +1,150 @@
+            paramMap.put(ParamsNames.EMAIL_SUBJECT, emailSubject);
+            paramMap.put(ParamsNames.EMAIL_CONTENT, message.getContent());
+            paramMap.put(ParamsNames.EMAIL_SENDER, emailSender);
+            paramMap.put(ParamsNames.EMAIL_RECEIVER, emailReceiver);
+            paramMap.put(ParamsNames.EMAIL_REPLY_TO_ADDRESS, emailReplyToAddress);
+            
+            TaskQueuesLogic taskQueueLogic = TaskQueuesLogic.inst();
+            taskQueueLogic.createAndAddDeferredTask(SystemParams.SEND_EMAIL_TASK_QUEUE,
+                    Const.ActionURIs.SEND_EMAIL_WORKER, paramMap, emailDelayTimer);
+        } catch (Exception e) {
+            log.severe("Error when adding email to task queue: " + e.getMessage() + "\n"
+                       + "Email sender: " + emailSender + "\n"
+                       + "Email receiver: " + emailReceiver + "\n"
+                       + "Email subject: " + emailSubject + "\n"
+                       + "Email reply to address: " + emailReplyToAddress);
+        }
+    }
+    
+    /**
+     * Sends the given {@code message} and generates a log report.
+     */
+    public void sendEmailWithLogging(EmailWrapper message) throws Exception {
+        sendEmail(message, true);
+    }
+    
+    /**
+     * Sends the given {@code message} without generating a log report.
+     */
+    public void sendEmailWithoutLogging(EmailWrapper message) throws Exception {
+        sendEmail(message, false);
+    }
+    
+    private void sendEmail(EmailWrapper message, boolean isWithLogging) throws Exception {
+        service.sendEmail(message);
+        if (isWithLogging) {
+            generateLogReport(message);
+        }
+    }
+    
+    private void generateLogReport(EmailWrapper message) {
+        try {
+            EmailLogEntry newEntry = new EmailLogEntry(message);
+            String emailLogInfo = newEntry.generateLogMessage();
+            log.info(emailLogInfo);
+        } catch (Exception e) {
+            log.severe("Failed to generate log for email: " + message.getInfoForLogging());
+        }
+    }
+    
+    /**
+     * Sends the given {@code errorReport}.
+     */
+    public void sendErrorReport(EmailWrapper errorReport) throws Exception {
+        sendEmailWithoutLogging(errorReport);
+        log.info("Sent crash report: " + errorReport.getInfoForLogging());
+    }
+    
+    /**
+     * Reports that a system {@code error} has occurred and the {@code errorReport} that is
+     * supposed to report it has failed to sent.<br>
+     * This method can be used when the usual error report sending fails to make sure that
+     * no stack traces are lost in the process.
+     * @param error the original error to be reported in {@code errorReport}
+     * @param errorReport the report that fails to send
+     * @param e the exception which causes {@code errorReport} to fail to send
+     */
+    public void reportErrorWithBackupChannel(Throwable error, EmailWrapper errorReport, Exception e) {
```

Comment:
You could've asked about this in the previous PR...
Anyway, it's just for a lack of better name of "a method that cannot fail because it doesn't throw any Exception" :p